### DISCUSSION TITLE
Problems initializing own constitutive model

### DISCUSSION THREAD

**Discussion by scja1991** [KratosMultiphysics/Kratos#11560] [Q&A] [answered]
I want to implement a hardening law for plasticity. For this I work with the application ConstitutiveModelsApplication. I replaced the deprecated Python commands and the material models already implemented in this application are now running. I added my plasticity model to the 'add_custom_constitutive_laws_to_python.cpp' file bute get the error AttributeError: 'module 'KratosMultiphysics.ConstitutiveModelsApplication' has no attribute 'VonMisesLinearElasticSwiftPlasticityModel''
My files are listet in 'CMakeLists.txt'.

I am using VisualCode on Windows 11 Pro and Pythonn 3.11.3 on a 64-bit machine.

Has someone a hint where I should look for my fault or what is the right way to find it?

**Comment by matekelemen** [comment]
Can you plz produce a diff on your `applications/ConstutitutiveLawsApplication/CMakeLists.txt` and 
`applications/ConstutitutiveLawsApplication/custom_python/add_custom_constitutive_laws_to_python.cpp` and post them here?

**Comment by scja1991** [comment]
Sure:

---------- applications/ConstitutiveModelsApplication/CMakeLists.txt ----------
@@ -55,6 +55,7 @@ set( KRATOS_CONSTITUTIVE_MODELS_APPLICATION_CORE

${CMAKE_CURRENT_SOURCE_DIR}/custom_models/plasticity_models/hardening_rules/modified_exponential_damage_hardening_rule.cpp
${CMAKE_CURRENT_SOURCE_DIR}/custom_models/plasticity_models/hardening_rules/cam_clay_hardening_rule.cpp
${CMAKE_CURRENT_SOURCE_DIR}/custom_models/plasticity_models/hardening_rules/gens_nova_hardening_rule.cpp
+${CMAKE_CURRENT_SOURCE_DIR}/custom_models/plasticity_models/hardening_rules/swift_hardening_rule.cpp
 
#utilities
${CMAKE_CURRENT_SOURCE_DIR}/custom_utilities/properties_layout.cpp

 applications/ConstitutiveModelsApplication/custom_python/add_custom_constitutive_laws_to_python.cpp 
@@ -56,6 +56,7 @@
 
 // Plasticity models
 #include "custom_models/plasticity_models/von_mises_linear_elastic_plasticity_model.hpp"
+#include "custom_models/plasticity_models/von_mises_linear_elastic_swift_plasticity_model.hpp"
 #include "custom_models/plasticity_models/von_mises_neo_hookean_plasticity_model.hpp"
 #include "custom_models/plasticity_models/simo_J2_plasticity_model.hpp"
 #include "custom_models/plasticity_models/simo_J2_thermo_plasticity_model.hpp"

**Reply by matekelemen** [reply, marked as answer]
Right, so far so good. However, it's not enough to include the header of your class, you have to define its bindings manually. You can take a look at the rest of `add_custom_constitutive_laws_to_python.cpp` for examples. You'll probably have to define your bindings like this:
```c++
    py::class_<VonMisesLinearElasticSwiftPlasticityModel, VonMisesLinearElasticSwiftPlasticityModel::Pointer, ConsitutiveLaw>(m, "VonMisesLinearElasticSwiftPlasticityModel")
        .def(py::init<>())
        ;
```

### Details

*This is just extra info if you want to know more about python bindings - it's not strictly necessary to understand it to solve your problem at hand.*

We're using [pybind](https://pybind11.readthedocs.io/en/latest/) to create python bindings for C++ classes. You'll need to be fairly comfortable with C++ to understand how pybind works, but in this case, this is all you need to know:
- you have to specify
   - the class you wish to bind `VonMisesLinearElasticSwiftPlasticityModel` - first template parameter to `py::class_`
   - the handle type of your class, in this case `VonMisesLinearElasticSwiftPlasticityModel::Pointer` - second template parameter
   - the base class of your bound class: `ConstitutiveLaw` (I'm assuming you derived from it) - third template parameter
- you also need to provide the name of your class separately `"VonMisesLinearElasticSwiftPlasticityModel"`
- you have to define bindings for each member function you wish to invoke from python - including the default constructor, which is what `py::init<>` is for. The bound class inherits the member functions of its base (`ConsitutiveLaw`) so you probably don't have to bind anything else, unless you have some custom stuff you want to reach from python.

Pybind uses a [fluent interface](https://en.wikipedia.org/wiki/Fluent_interface) to define its bindings, so you can chain `py::class_::def`s to define bindings for multiple member functions.

**Comment by matekelemen** [comment]
@scja1991 any luck?

**Reply by scja1991** [reply]
Yes, this was the hint I needed!

**Comment by AlejandroCornejo** [comment]
I do not know how is maintining the `ConstitutiveModelsApp`. I am the main developer and maintainer of the `ConstitutiveLawsApp`. IMO you should check mine since it is currently maintained in the master.

**Reply by scja1991** [reply]
I took a look at your app. But I didn't fully understand where to implement more hardening rules there.

**Reply by AlejandroCornejo** [reply]
The hardening rules for plasticity are implemented in here https://github.com/KratosMultiphysics/Kratos/blob/b652b8ebca7d1cc2d7a5fd5d1aee7dd560fe1ba6/applications/ConstitutiveLawsApplication/custom_constitutive/auxiliary_files/cl_integrators/generic_cl_integrator_plasticity.h#L446
